Changed AI.minMax return value and disabled A-B pruning

vincentjguo opened the pull request:
AI.minMax now returns a TreeMap with a column as its key and the generated score as its value. This should make the function AI.bestColumn more accurate in terms of associating the generated scores with a column.

Alpha-Beta pruning is efficient but results in the AI making bad moves. This function would probably be disabled in the final release. An adapted version of A-B pruning is used if a player wins.

vincentjguo committed:
Changed AI.minMax return value and disabled A-B pruning

AI.minMax now returns a TreeMap with a column as its key and the generated score as its value. This should make the function AI.bestColumn more accurate in terms of associating the generated scores with a column.

Alpha-Beta pruning is efficient but results in the AI making bad moves. This function would probably be disabled in the final release. An adapted version of A-B pruning is used if a player wins.

Signed-off-by: Vincent Guo <vincentguo04@gmail.com>

vincentjguo commented:
This PR might fix #46 as AB pruning might be the cause of this issue.

potatoeggy commented:
Why does Alpha-Beta pruning cause the AI to make bad moves? Isn't it supposed to remove nodes that are guaranteed to be worse than a current alternative?

vincentjguo commented:
> Why does Alpha-Beta pruning cause the AI to make bad moves? Isn't it supposed to remove nodes that are guaranteed to be worse than a current alternative?

Well it's probably because I'm just bad at implementing.

potatoeggy committed:
menus: disable player choice for player 2

potatoeggy committed:
ai: mark todo for code cleanup

potatoeggy committed:
game: reduce general difficulty for performance

potatoeggy commented:
Performance issues have been worked around by reducing depths to 1, 2, and 3 for AI-controlled difficulties.

E: This means that we only hit severe lag at "Hard" difficulty and higher.

potatoeggy commented on src/AI.java:
```diff
@@ -221,27 +221,23 @@ static int bestColumn(Board board, int difficulty) {
 		}
-		ArrayList<Integer> bestRows = minMax(board, depth, 1, Integer.MIN_VALUE, Integer.MAX_VALUE); // grab value from big algorithm
-		System.out.println(Arrays.toString(bestRows.toArray()));
+		TreeMap<Integer, Integer> bestRows = minMax(board, depth, 1, Integer.MIN_VALUE, Integer.MAX_VALUE); // grab value from big algorithm
```
What's the reason for using a TreeMap versus a HashMap?

vincentjguo replied:
Easier to iterate through, but a hashmap works too and it should be faster now that I think about it.

potatoeggy replied:
HashMap is O(1) for inserting, grabbing, and removing on average, while TreeMap is O(log(n))

potatoeggy commented on src/AI.java:
```diff
@@ -151,55 +149,57 @@ else if (curr.color < 3 && curr.rowSize != 3) {
-					for (int j : value) {
-						max = Math.max(max, j);
+					int max = Integer.MIN_VALUE, index = -1;
+					for (Map.Entry<Integer, Integer> j : value.entrySet()) {
```
If possible, maybe try to use TreeMap's ability to sort things to grab the largest key. That probably needs a lot of refactoring, though.

vincentjguo replied:
[Here's how to sort by value](https://stackoverflow.com/questions/2864840/treemap-sort-by-value). Kind of difficult to implement

vincentjguo commented:
Are we still using javadoc?

potatoeggy commented:
We don't have to. I just added it into the GUI because I ran out of stuff to do.

On Jun 9 2020, at 2:39 pm, Vincent Guo <notifications@github.com> wrote:
>
>
> Are we still using javadoc?
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub (https://github.com/potatoeggy/connectfour/pull/53#issuecomment-641499050), or unsubscribe (https://github.com/notifications/unsubscribe-auth/AGADGXVSRDVNMKD5STHUWA3RVZ6X3ANCNFSM4NZTKNZA).
>

vincentjguo committed:
Added some javaDoc documentation and improved TreeMap implementation.

TreeMap now uses generated score as key so it is able to passively sort it.
More documentation as per issue #2's request.

Signed-off-by: Vincent Guo <vincentguo04@gmail.com>

vincentjguo committed:
Added some javaDoc documentation and improved TreeMap implementation.

TreeMap now uses generated score as key so it is able to passively sort it.
More documentation as per issue #2's request.

Signed-off-by: Vincent Guo <vincentguo04@gmail.com>

vincentjguo commented:
Same i guess

potatoeggy commented:
We should get rid of debug statements if they aren't needed anymore

vincentjguo committed:
Cleaned up the code and removed debug statements

Signed-off-by: Vincent Guo <vincentguo04@gmail.com>
